sync15-adapter/src/token.rs

```diff
@@ -131,3 +143,185 @@ impl TokenserverClient {
+                    }
+                    _ => {
+                        TokenState::Failed(e, previous_endpoint)
+                    }
+                }
+            }
+        }
+    }
+
+    // Given the state we are currently in, return a new current state.
+    // Returns None if the current state should be used (eg, if we are
+    // holding a token that remains valid) or Some() if the state has changed
+    // (which may have changed to a state with a token or an error state)
+    fn advance_state(&self, request_client: &Client, state: &TokenState) -> Option<TokenState> {
+        match state {
+            TokenState::NoToken => {
+                Some(self.fetch_token(request_client, None))
+            },
+            TokenState::Failed(_, existing_endpoint) => {
+                Some(self.fetch_token(request_client, existing_endpoint.clone()))
+            },
+            TokenState::Token(existing_context) => {
+                if existing_context.is_valid() {
+                    None
+                } else {
+                    Some(self.fetch_token(request_client, Some(existing_context.token.api_endpoint.clone())))
+                }
+            },
+            TokenState::Backoff(ref until, ref existing_endpoint) => {
+                if let Ok(remaining) = until.duration_since(SystemTime::now()) {
+                    debug!("enforcing existing backoff - {:?} remains", remaining);
+                    None
+                } else {
+                    // backoff period is over
+                    Some(self.fetch_token(request_client, existing_endpoint.clone()))
+                }
+            },
+            TokenState::NodeReassigned => {
+                // We never leave this state.
+                None
+            }
+        }
+    }
+
+    fn with_token<T, F>(&self, request_client: &Client, func: F) -> Result<T>
+            where F: FnOnce(&TokenContext) -> Result<T> {
+
+        // first get a mutable ref to our existing state, advance to the
+        // state we will use, then re-stash that state for next time.
+        let state: &mut TokenState = &mut self.current_state.borrow_mut();
+        match self.advance_state(request_client, state) {
+            Some(new_state) => *state = new_state,
+            None => ()
+        }
+
+        // Now re-fetch the state we should use for this call - if it's
+        // anything other than TokenState::Token we will fail.
+        match state {
+            TokenState::NoToken => {
+                // it should be impossible to get here.
+                panic!("Can't be in NoToken state after advancing");
+            }
+            TokenState::Token(ref token_context) => {
+                // make the call.
+                func(token_context)
+            }
+            TokenState::Failed(_, _) => {
+                // XXX - todo - work out how to return the value???
+                bail!(error::unexpected("todo - work out how to return the actual error?"))
```

Comment:
`TokenState::Failed(e, _) => return Err(e)` or something like that doesn't work?